Index: ejabberd/src/ejabberd_sm.erl
===================================================================
--- ejabberd.orig/src/ejabberd_sm.erl
+++ ejabberd/src/ejabberd_sm.erl
@@ -555,7 +555,7 @@
 					  From, To, IQ);
 		[] ->
 		    Err = jlib:make_error_reply(
-			    Packet, ?ERR_SERVICE_UNAVAILABLE),
+			    Packet, ?ERR_FEATURE_NOT_IMPLEMENTED),
 		    ejabberd_router:route(To, From, Err)
 	    end;
 	reply ->
