 1) Simple typo fix from the ejabberd author.

Index: ejabberd/src/ejabberd_ctl.erl
===================================================================
--- ejabberd.orig/src/web/ejabberd_http_poll.erl
+++ ejabberd/src/web/ejabberd_http_poll.erl
@@ -208,7 +208,7 @@
 	    Receiver ! {tcp, {http_poll, self()}, list_to_binary(Input)},
 	    {reply, ok, StateName, StateData#state{input = "",
 						   waiting_input = false,
-						   last_receiver = From
+						   last_receiver = Receiver
 						  }}
     end;
 
