From: Evgeniy Khramtsov <ekhramtsov@process-one.net>
Subject: Fix getting vCards with a PHOTO element via mod_vcard_ldap (EJAB-1258)
Origin: upstream, https://git.process-one.net/ejabberd/mainline/commit/3024bb0cbf359f3e14b5386ed14ecfb168e8a267
Bug: https://support.process-one.net/browse/EJAB-1258
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/ejabberd/+bug/613854
--- a/src/eldap/eldap_filter.erl
+++ b/src/eldap/eldap_filter.erl
@@ -171,7 +171,7 @@
     end.
 
 replace_amps(String) ->
-    lists:map(
+    lists:flatmap(
       fun($&) -> "\\&";
-	 (Chr) -> Chr
+	 (Chr) -> [Chr]
       end, String).
