Index: pkg-ejabberd/src/ejabberd_app.erl
===================================================================
--- pkg-ejabberd.orig/src/ejabberd_app.erl	2010-05-17 21:52:29.000000000 +0400
+++ pkg-ejabberd/src/ejabberd_app.erl	2010-05-17 21:52:44.000000000 +0400
@@ -85,7 +85,7 @@
 stop(_State) ->
     ?INFO_MSG("ejabberd ~s is stopped in the node ~p", [?VERSION, node()]),
     delete_pid_file(),
-    ejabberd_debug:stop(),
+    %ejabberd_debug:stop(),
     ok.
 
 
