From 674afcdf4eb628a4d9517c819c48a38137ca55d3 Mon Sep 17 00:00:00 2001
From: Badlop <badlop@process-one.net>
Date: Thu, 8 Dec 2011 12:50:40 +0100
Subject: [PATCH 2/4] Add callback function print_state/1 for behavior p1_fsm

---
 src/ejabberd_s2s_in.erl |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

--- a/src/ejabberd_s2s_in.erl
+++ b/src/ejabberd_s2s_in.erl
@@ -44,6 +44,7 @@
 	 handle_sync_event/4,
 	 code_change/4,
 	 handle_info/3,
+	 print_state/1,
 	 terminate/3]).
 
 -include("ejabberd.hrl").
@@ -668,6 +669,13 @@
     Shaper = acl:match_rule(Host, StateData#state.shaper, JID),
     (StateData#state.sockmod):change_shaper(StateData#state.socket, Shaper).
 
+%%----------------------------------------------------------------------
+%% Func: print_state/1
+%% Purpose: Prepare the state to be printed on error log
+%% Returns: State to print
+%%----------------------------------------------------------------------
+print_state(State) ->
+    State.
 
 new_id() ->
     randoms:get_string().
