Index: ejabberd/rebar.config.script
===================================================================
--- ejabberd.orig/rebar.config.script	2014-05-15 08:03:26.962505041 +0200
+++ ejabberd/rebar.config.script	2014-05-15 08:03:52.071081244 +0200
@@ -127,9 +70,8 @@
           {keep_build_info, true},
           {ct_extra_params, "-include "
            ++ filename:join([Cwd, "tools"]) ++ " "
-           ++ filename:join([Cwd, "deps", "p1_xml", "include"])},
-          {post_hooks, PostHooks ++ CfgPostHooks},
-          {deps, Deps ++ CfgDeps}],
+           ++ filename:join([Cwd, "deps", "p1_xml", "include"])}
+	 ],
 %%io:format("ejabberd configuration:~n  ~p~n", [Config]),
 Config.
 
